From 21d98e14d5f35e7fce1a86bb7dc4b2ee767602fc Mon Sep 17 00:00:00 2001
From: Thomas Epperson <thomas.epperson@gmail.com>
Date: Sat, 15 Sep 2018 11:05:01 -0400
Subject: [PATCH 3/3] Revert some changes implemented to make it compile. Pass
 --divide to the gcc assembler so it doesn't treat / as the start of a
 comment.

---
 Makefile           | 9 ++++++---
 boot/BootStartup.S | 5 ++---
 boot_xbe/Makefile  | 2 +-
 3 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/Makefile b/Makefile
index c24705b..c915ec2 100644
--- a/Makefile
+++ b/Makefile
@@ -11,10 +11,10 @@ INCLUDE += -I$(TOPDIR)/grub -I$(TOPDIR)/include -I$(TOPDIR)/ -I./ -I$(TOPDIR)/fs
 	-I$(TOPDIR)/lib/jpeg/ -I$(TOPDIR)/menu/actions -I$(TOPDIR)/menu/textmenu -I$(TOPDIR)/menu/iconmenu
 
 #These are intended to be non-overridable.
-CROM_CFLAGS=$(INCLUDE)
+CROM_CFLAGS=$(INCLUDE) -Wa,--divide
 
 #You can override these if you wish.
-CFLAGS= -Os -march=pentium -m32 -Werror -Wstrict-prototypes -Wreturn-type -fomit-frame-pointer  -DIPv4 -fpack-struct -ffreestanding -fno-PIC
+CFLAGS= -Os -march=pentium -m32 -Werror -Wstrict-prototypes -Wreturn-type -fomit-frame-pointer  -DIPv4 -fpack-struct -ffreestanding -fno-PIC -Wa,--divide
 
 # add the option for gcc 4.2 only, again, non-overridable
 ifeq ($(GCC_4.2), 1)
@@ -34,7 +34,7 @@ ifeq ($(ETHERBOOT), yes)
 ETH_SUBDIRS = etherboot
 CROM_CFLAGS	+= -DETHERBOOT
 ETH_INCLUDE = 	-I$(TOPDIR)/etherboot/include -I$(TOPDIR)/etherboot/arch/i386/include	
-ETH_CFLAGS  = 	-O2 -march=pentium -mtune=pentium -Werror $(ETH_INCLUDE) -Wstrict-prototypes -fomit-frame-pointer -pipe -Ui386
+ETH_CFLAGS  = 	-O2 -march=pentium -mtune=pentium -Werror $(ETH_INCLUDE) -Wstrict-prototypes -fomit-frame-pointer -pipe -Ui386 -Wa,--divide
 endif
 
 LDFLAGS-ROM     = -s -S -T $(TOPDIR)/scripts/ldscript-crom.ld
@@ -232,3 +232,6 @@ imagecompress: obj/image-crom.bin bin/imagebld
 	bin/imagebld -xbe xbe/xromwell.xbe obj/image-crom.bin
 	bin/imagebld -vml boot_vml/disk/vmlboot obj/image-crom.bin 
 
+%.o     : %.S
+	${CC} -DASSEMBLER ${CFLAGS} -Wa,--divide -o $@ -c $<
+
diff --git a/boot/BootStartup.S b/boot/BootStartup.S
index 4aac811..326e46b 100644
--- a/boot/BootStartup.S
+++ b/boot/BootStartup.S
@@ -102,8 +102,7 @@ start_linux:
 	// Copy the GDT to its final location
 	movl $GDT_LOC, %edi
 	movl $tableGdt, %esi
-	movl $(tableGdtEnd-tableGdt), %ecx
-	shrl $2, %ecx
+	movl $(tableGdtEnd-tableGdt)/4, %ecx
 	rep movsl
 
 	// Load the new GDT
@@ -139,7 +138,7 @@ reload_cs:
         
     // We clear the IDT in RAM   (IDT located @  0xb0000 )
 	xor %eax, %eax
-	movl $0x1400, %ecx
+	movl $0x5000/4, %ecx
 	movl $IDT_LOC, %edi
 	rep stosl
 	wbinvd
diff --git a/boot_xbe/Makefile b/boot_xbe/Makefile
index b5e203b..0889023 100644
--- a/boot_xbe/Makefile
+++ b/boot_xbe/Makefile
@@ -4,5 +4,5 @@ O_TARGET := xbeboot.o
 xbeboot.o:
 
 %.o     : %.S
-	${CC} -DASSEMBLER ${CFLAGS} -o $@ -c $<
+	${CC} -DASSEMBLER ${CFLAGS} -Wa,--divide -o $@ -c $<
 
-- 
1.9.1

